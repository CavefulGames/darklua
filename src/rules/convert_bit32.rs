use crate::nodes::{BinaryExpression, Block, Expression, LocalAssignStatement, BinaryOperator};
use crate::process::{DefaultVisitor, Evaluator, NodeProcessor, NodeVisitor};
use crate::rules::{
    Context, FlawlessRule, RuleConfiguration, RuleConfigurationError, RuleProperties,
};

use super::verify_no_rule_properties;

#[derive(Default)]
struct Processor {}

impl NodeProcessor for Processor {
    fn process_binary_expression(&mut self, bin_exp: &mut crate::nodes::BinaryExpression) {
        let left = bin_exp.left();
        let right = bin_exp.right();
        *bin_exp = BinaryExpression::new(
            BinaryOperator::DoubleGreaterThan,
            left.clone(),
            right.clone()
        );
    }
}

pub const CONVERT_BIT32_RULE_NAME: &str = "convert_bit32";

/// A rule that removes trailing `nil` in local assignments.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct ConvertBit32 {}

impl FlawlessRule for ConvertBit32 {
    fn flawless_process(&self, block: &mut Block, _: &Context) {
        let mut processor = Processor::default();
        DefaultVisitor::visit_block(block, &mut processor);
    }
}

impl RuleConfiguration for ConvertBit32 {
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        verify_no_rule_properties(&properties)?;

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        CONVERT_BIT32_RULE_NAME
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        RuleProperties::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::rules::Rule;

    use insta::assert_json_snapshot;

    fn new_rule() -> ConvertBit32 {
        ConvertBit32::default()
    }

    #[test]
    fn serialize_default_rule() {
        let rule: Box<dyn Rule> = Box::new(new_rule());

        assert_json_snapshot!("default_convert_bit32", rule);
    }

    #[test]
    fn configure_with_extra_field_error() {
        let result = json5::from_str::<Box<dyn Rule>>(
            r#"{
            rule: 'convert_bit32',
            prop: "something",
        }"#,
        );
        pretty_assertions::assert_eq!(result.unwrap_err().to_string(), "unexpected field 'prop'");
    }
}
