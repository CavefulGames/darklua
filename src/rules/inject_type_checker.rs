use std::collections::HashMap;

use regex::Regex;

use super::runtime_identifier::RuntimeIdentifierBuilder;
use super::{
    Context, Rule, RuleConfiguration, RuleConfigurationError, RuleProcessResult,
    RuleProperties,
};
use crate::nodes::{
    Arguments, BinaryExpression, BinaryOperator, Block, Expression, FunctionCall, FunctionStatement, Identifier, IfBranch, IfStatement, LastStatement, LocalFunctionStatement, Prefix, ReturnStatement, Statement, StringExpression, Type, TypedIdentifier
};
use crate::process::{DefaultVisitor, NodeProcessor, NodeVisitor};

struct ErrorCallBuilder {
    name: String,
    message_format: String,
}

impl ErrorCallBuilder {
    fn build(
        &self,
        index: usize,
        name: &str,
        expected_type: &str,
        actual_type: &str,
    ) -> FunctionCall {
        let index = index.to_string();
        let mut vars = HashMap::new();
        vars.insert("index".to_owned(), index.as_str());
        vars.insert("name".to_owned(), name);
        vars.insert("expected_type".to_owned(), expected_type);
        vars.insert("actual_type".to_owned(), actual_type);

        FunctionCall::new(
            Prefix::Identifier(Identifier::new(self.name.as_str())),
            Arguments::String(StringExpression::from_value(self.message_format.as_str())),
            None
        )
    }
}

trait TypeCheckerBuilder: std::fmt::Debug {
    fn build(&self, parameters: &Vec<TypedIdentifier>) -> (Statement);
}

struct TypeCheckerConfig {
    ignore_comment_pattern: Option<Regex>,
    type_variable_name: String,
    error_call_builder: ErrorCallBuilder,
    return_errors: bool,
    indexed_functions_only: bool,
}

fn get_possible_type_names(t: &Type) -> Option<Vec<(String, bool)>> {
    match t {
        Type::Array(_) => {
            Some(vec![("table".to_owned(), false)])
        }
        Type::False(_) => {
            Some(vec![("boolean".to_owned(), false)])
        }
        Type::Function(_) => {
            Some(vec![("function".to_owned(), false)])
        }
        Type::Intersection(intersection_type) => {
            let mut type_names: Vec<(String, bool)> = Vec::new();
            let left_type = intersection_type.get_left();
            if let Some(names) = get_possible_type_names(left_type) {
                for name in names {
                    type_names.push(name);
                }
            }

            let right_type = intersection_type.get_right();
            if let Some(names) = get_possible_type_names(right_type) {
                for name in names {
                    type_names.push(name);
                }
            }

            let first = &type_names[0];

            if type_names.iter().all(|x| x == first) {
                Some(vec![first.clone()])
            } else {
                None
            }
        }
        Type::Name(name) => {
            let name = name.get_type_name().get_name().as_str();
            match name {
                "nil" | "string" | "number" | "boolean" | "thread" | "buffer" => {
                    Some(vec![(name.to_owned(), false)])
                },
                _ => {
                    None
                }
            }
        }
        Type::Nil(_) => {
            Some(vec![("nil".to_owned(), false)])
        }
        Type::Optional(optional_type) => {
            let inner_type = optional_type.get_inner_type();
            if let Some(names) = get_possible_type_names(inner_type) {
                Some(names.iter().map(|(name, _)| (name.clone(), true)).collect())
            } else {
                None
            }
        }
        Type::Parenthese(parenthese_type) => {
            let inner_type = parenthese_type.get_inner_type();
            get_possible_type_names(inner_type)
        }
        Type::String(_) => {
            Some(vec![("nil".to_owned(), false)])
        }
        Type::Table(_) => {
            Some(vec![("table".to_owned(), false)])
        }
        Type::True(_) => {
            Some(vec![("boolean".to_owned(), false)])
        }
        Type::Union(union_type) => {
            let mut type_names: Vec<(String, bool)> = Vec::new();
            let left_type = union_type.get_left();
            if let Some(names) = get_possible_type_names(left_type) {
                for name in names {
                    type_names.push(name);
                }
            }

            let right_type = union_type.get_right();
            if let Some(names) = get_possible_type_names(right_type) {
                for name in names {
                    type_names.push(name);
                }
            }

            if type_names.is_empty() {
                None
            } else {
                Some(type_names)
            }
        }
        _ => {
            None
        }
    }
}

struct NonStrict {
    config: TypeCheckerConfig,
}

impl TypeCheckerBuilder for NonStrict {
    fn build(&self, parameters: &Vec<TypedIdentifier>) -> Statement {

    }
}

struct Strict {
    config: TypeCheckerConfig,
}

impl TypeCheckerBuilder for Strict {
    fn build(&self, parameters: &Vec<TypedIdentifier>) -> Statement {
        let branches: Vec<IfBranch> = Vec::new();
        parameters
            .iter()
            .filter_map(|param| param.get_type())
            .filter_map(|t| get_possible_type_names(t))
            .for_each(|t| { // condition: type(x) ~= "x"
                let condition = BinaryExpression::new(BinaryOperator::NotEqual, left, right);
                let if_branch = IfBranch::new(condition, block);
                branches.push(if_branch);
            });

        IfStatement::new(branches, None).into()
    }
}

struct FunctionProcessor {
    type_checker_builder: Box<dyn TypeCheckerBuilder>,
}

impl NodeProcessor for FunctionProcessor {
    fn process_statement(&mut self, stmt: &mut Statement) {
        match stmt {
            Statement::Function(func_stmt) => {
                if !func_stmt.has_parameters() {
                    return;
                }
                let stmt = self.type_checker_builder.build(func_stmt.get_parameters());
                let block = func_stmt.mutate_block();
                block.insert_statement(0, stmt);
            }
            Statement::LocalFunction(local_func_stmt) => {
                if !local_func_stmt.has_parameters() {
                    return;
                }
                let stmt = self.type_checker_builder.build(local_func_stmt.get_parameters());
                let block = local_func_stmt.mutate_block();
                block.insert_statement(0, stmt);
            }
            _ => {}
        }
    }
}

pub const INJECT_TYPE_CHECKER_RULE_NAME: &str = "inject_type_checker";

/// A rule that injects type checkers in function statements.
#[derive(Debug, PartialEq)]
pub struct InjectTypeChecker {
    error_call: String,
    error_message_format: String,
    return_errors: bool,
    ignore_comment_pattern: Option<String>,
    indexed_functions_only: bool,
    strict: bool,
    runtime_identifier_format: String,
    ignore_local_functions: bool,
}

impl Default for InjectTypeChecker {
    fn default() -> Self {
        Self {
            error_call: "error".to_owned(),
            error_message_format: "invalid argument #{index} to '{name}' ({expected_type} expected, got {actual_type})".to_owned(),
            return_errors: false,
            ignore_comment_pattern: None,
            indexed_functions_only: false,
            strict: false,
            runtime_identifier_format: "__DARKLUA_INJECT_TYPE_CHECKER_{name}{hash}".to_owned(),
            ignore_local_functions: false,
        }
    }
}

impl Rule for InjectTypeChecker {
    fn process(&self, block: &mut Block, context: &Context) -> RuleProcessResult {
        let pattern = if let Some(pattern) = &self.ignore_comment_pattern {
            Some(Regex::new(pattern.as_str()).map_err(|err| err.to_string())?)
        } else {
            None
        };

        let identifier_builder = RuntimeIdentifierBuilder::new(
            self.runtime_identifier_format.as_str(),
            context.original_code.as_bytes(),
            None,
        )?;

        let error_call_builder = ErrorCallBuilder {
            name: self.error_call.clone(),
            message_format: self.error_message_format.clone(),
        };

        let config = TypeCheckerConfig {
            ignore_comment_pattern: pattern,
            type_variable_name: identifier_builder.build("t")?,
            error_call_builder,
            return_errors: self.return_errors,
            indexed_functions_only: self.indexed_functions_only,
        };

        let type_checker_builder = Box::new(Strict { config });

        let mut function_processor = FunctionProcessor { type_checker_builder };

        DefaultVisitor::visit_block(block, &mut function_processor);

        Ok(())
    }
}

impl RuleConfiguration for InjectTypeChecker {
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        for (key, value) in properties {
            match key.as_str() {
                "error_call" => {
                    self.error_call = value.expect_string(&key)?;
                }
                "error_message_format" => {
                    self.error_message_format = value.expect_string(&key)?;
                }
                "return_errors" => {
                    self.return_errors = value.expect_bool(&key)?;
                }
                "ignore_comment_pattern" => {
                    self.ignore_comment_pattern = Some(value.expect_string(&key)?);
                }
                "indexed_functions_only" => {
                    self.indexed_functions_only = value.expect_bool(&key)?;
                }
                "strict" => {
                    self.strict = value.expect_bool(&key)?;
                }
                "ignore_local_functions" => {
                    self.ignore_local_functions = value.expect_bool(&key)?;
                }
                "runtime_identifier" => {
                    self.runtime_identifier_format = value.expect_string(&key)?;
                }
                _ => return Err(RuleConfigurationError::UnexpectedProperty(key)),
            }
        }

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        INJECT_TYPE_CHECKER_RULE_NAME
    }

    fn serialize_to_properties(&self) -> super::RuleProperties {
        RuleProperties::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::rules::Rule;

    use insta::assert_json_snapshot;

    fn new_rule() -> InjectTypeChecker {
        InjectTypeChecker::default()
    }

    #[test]
    fn serialize_default_rule() {
        let rule: Box<dyn Rule> = Box::new(new_rule());

        assert_json_snapshot!("default_inject_type_checker", rule);
    }

    #[test]
    fn configure_with_extra_field_error() {
        let result = json5::from_str::<Box<dyn Rule>>(
            r#"{
            rule: 'inject_type_checker',
            error_call: 'error',
            error_message_format: 'foo',
            return_errors: false,
            indexed_functions_only: false,
            strict: true,
            ignore_local_functions: false,
            runtime_identifier: "{name}",
            prop: "something",
        }"#,
        );
        pretty_assertions::assert_eq!(result.unwrap_err().to_string(), "unexpected field 'prop'");
    }
}
