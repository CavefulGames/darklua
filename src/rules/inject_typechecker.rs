use log::error;
use serde::{Deserialize, Serialize};

use crate::process::{DefaultVisitor, NodeProcessor, NodeVisitor};
use crate::nodes::{Arguments, BinaryExpression, BinaryOperator, Block, Expression, FunctionCall, FunctionStatement, Identifier, IfBranch, IfStatement, LastStatement, Prefix, ReturnStatement, Statement, StringExpression, Type};
use super::{Context, FlawlessRule, RuleConfiguration, RuleConfigurationError, RuleProperties};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields, rename_all = "snake_case", tag = "name")]
pub struct TypecheckerType {
    identifier: String,
    method: Option<String>,
    call: Option<String>
}

#[derive(Debug)]
struct Processor {
    use_typeof: String,
    return_errors: bool,
    error_message: String,
    error_call: String,
    types: Option<Vec<TypecheckerType>>,
}

impl Processor {
    pub fn new(use_typeof: String, return_errors: bool, error_message: String, error_call: String, types: Option<Vec<TypecheckerType>>) -> Self {
        Self {
            use_typeof: use_typeof,
            return_errors: return_errors,
            error_message: error_message,
            error_call: error_call,
            types: types,
        }
    }
}

impl NodeProcessor for Processor {

    fn process_function_statement(&mut self, fn_stmt: &mut FunctionStatement) {
        if !fn_stmt.has_parameters() {
            return;
        }

        // inject_typechecker 옵션 추가해야할거

        // 특정 주석 조건
        // 인덱스된 함수에만 추가하는가? 옵션

        // types: [
        //   {
        //       identifier: "Signal", // 만약 함수 파라매터 타입이 이거인가
        //       method: "is", // signal:is()로 체크함
        //       call: "type" // type(signal)로 체크함
        //   }
        // ]

        // 문자열, 숫자 호환성 고려 & runtime variable

        let mut if_stmts: Vec<IfBranch> = Vec::new();
        let mut index = 0;
        for stmt in fn_stmt.get_parameters() {
            index += 1;
            if let Some(t) = stmt.get_type() { 
                let type_name: Option<&Identifier> = if let Type::Name(t_name) = t { 
                    Some(t_name.get_type_name()) 
                } else {
                    None
                };
              
                if let Some(type_name) = type_name {
                    let stmt_expression = Expression::identifier(stmt.get_identifier().clone());
                    let stmt_argument = Arguments::with_argument(Arguments::default(), stmt_expression);
                    let stmt_type_of = FunctionCall::new(Prefix::from_name(self.use_typeof.as_str()), stmt_argument.clone(), None);
                    let stmt_string_expression = StringExpression::from_value(type_name.get_name());

                    let error_message = self.error_message.as_str();
                    let format_string_count = error_message.matches("{get_type}").count();
                    let mut format_str = str::replace(error_message, "{argument}", index.to_string().as_str());
                    format_str = str::replace(format_str.as_str(), "{func_name}", fn_stmt.get_name().get_name().clone().into_name().as_str());
                    format_str = str::replace(format_str.as_str(), "{original_type}", type_name.get_name());
                    format_str = str::replace(format_str.as_str(), "{get_type}", "%s");

                    let mut format_string_argument = Arguments::String(StringExpression::from_value(format_str));
                    let format_typeof_call = stmt_type_of.clone();

                    for _ in 1..format_string_count+1 {
                        format_string_argument = format_string_argument.clone().with_argument(Expression::Call(Box::new(format_typeof_call.clone())));
                    }

                    let call_expression = Expression::Call(Box::new(FunctionCall::new(Prefix::from_name("string.format"), format_string_argument , None)));
                    let call_argument = Arguments::with_argument(Arguments::default(), call_expression);
                    
                    let return_error = FunctionCall::new(Prefix::from_name("error"), call_argument, None);
                    let mut error_vec: Vec<Statement> = Vec::new();
                    let mut option_last_statement: Option<LastStatement> = None;
                    
                    if self.error_call != "" {
                        error_vec.push(Statement::Call(FunctionCall::new(Prefix::from_name(self.error_call.as_str()), Arguments::default(), None)));
                    }

                    if !self.return_errors {
                        error_vec.push(Statement::Call(return_error.clone()));
                    } else {
                        let last_statement = ReturnStatement::one(Expression::Call(Box::new(return_error))).into();
                        option_last_statement = Some(last_statement);
                    }
                    
                    if self.types != None {
                        if let Some(vec_typechecker) = &self.types {
                            for typechcker_type in vec_typechecker {
                                if &typechcker_type.identifier != type_name.get_name() {
                                    let new_block = Block::new(error_vec.clone(), option_last_statement.clone());
                                    if_stmts.push(IfBranch::new(BinaryExpression::new(BinaryOperator::NotEqual, stmt_type_of.clone(), stmt_string_expression.clone()), new_block));
                                    continue;
                                }
                                // let mut use_method: bool = false;
                                if &typechcker_type.call != &None {
                                    if &typechcker_type.method != &None {
                                        error!("In the types setting, only one of call and method can be used.")
                                    } else {
                                        // use call
                                        if let Some(call) = typechcker_type.call.clone() {
                                            let identifier = typechcker_type.identifier.clone();
                                            let fn_call = FunctionCall::new(Prefix::from_name(call), stmt_argument.clone(), None);

                                            let new_block = Block::new(error_vec.clone(), option_last_statement.clone());
                                            if_stmts.push(IfBranch::new(BinaryExpression::new(BinaryOperator::NotEqual, fn_call, Expression::String(StringExpression::from_value(identifier))), new_block));
                                        }
                                    }
                                } else if &typechcker_type.method != &None {
                                    // use method
                                    if let Some(method) = typechcker_type.method.clone() {
                                        let call = FunctionCall::new(Prefix::from_name(format!(" not {}", method)), stmt_argument.clone(), None);

                                        let new_block = Block::new(error_vec.clone(), option_last_statement.clone());
                                        if_stmts.push(IfBranch::new(Expression::Call(Box::new(call)), new_block));
                                    }
                                } else {
                                    error!("One of method and call must have a value.")
                                }
                            }
                        }
                    } else {
                        let new_block = Block::new(error_vec.clone(), option_last_statement.clone());

                        if_stmts.push(IfBranch::new( BinaryExpression::new(BinaryOperator::NotEqual, stmt_type_of, stmt_string_expression.clone()), new_block));
                    }
                }
            }
        };

        fn_stmt.mutate_block().insert_statement(0, IfStatement::new(if_stmts, None));
    }
}

pub const INJECT_TYPECHECKER_RULE_NAME: &str = "inject_typechecker";

#[derive(Debug, PartialEq)]
pub struct InjectTypechecker {
    use_typeof: String,
    return_errors: bool,
    error_message: String,
    error_call: String,
    types: Option<Vec<TypecheckerType>>,
}

impl Default for InjectTypechecker {
    fn default() -> Self {
        Self {
            use_typeof: "typeof".to_string(),
            return_errors: false,
            error_message: "invalid argument #{argument} to '{func_name}' ({original_type} expected, got {get_type})".to_string(),
            error_call: "".to_string(),
            types: None,
        }
    }
}

impl FlawlessRule for InjectTypechecker {
    fn flawless_process(&self, block: &mut Block, _: &Context) {
        let mut processor = Processor::new(self.use_typeof.clone(), self.return_errors, self.error_message.clone(), self.error_call.clone(), self.types.clone()) ;
        DefaultVisitor::visit_block(block, &mut processor);
    }
}

impl RuleConfiguration for InjectTypechecker {
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {

        for (key, value) in properties {
            match key.as_str() {
                "use_typeof" => {
                    self.use_typeof = if value.expect_bool(&key)? {
                        "typeof".to_string()
                    } else {
                        "type".to_string()
                    };
                },
                "return_errors" => {
                    self.return_errors = value.expect_bool(&key)?;
                },   
                "error_message" => {
                    
                    self.error_message = value.expect_string(&key)?;
                },
                "error_call" => {
                    self.error_call = value.expect_string(&key)?;
                },
                "types" => {
                    self.types = Some(value.expect_inject_typechecker_types(&key)?);
                },
                _ => return Err(RuleConfigurationError::UnexpectedProperty(key)),
            }
        }

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        INJECT_TYPECHECKER_RULE_NAME
    }

    fn serialize_to_properties(&self) -> super::RuleProperties {
        RuleProperties::new()
    }

}
