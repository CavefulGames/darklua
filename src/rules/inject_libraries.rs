use serde::{Deserialize, Serialize};
use std::ffi::OsString;
use std::fs;
use std::path::{Path, PathBuf};
use std::str::FromStr;

use crate::nodes::{
    Arguments, Block, FunctionCall, LocalAssignStatement, Prefix, StringExpression, TypedIdentifier,
};
use crate::rules::{Context, RuleConfiguration, RuleConfigurationError, RuleProperties};

use super::require::PathRequireMode;
use super::{verify_required_properties, RequireMode, Rule, RuleProcessResult};

use blake3;
use hex;
use path_slash::PathBufExt as _;
use pathdiff::diff_paths;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields, rename_all = "snake_case")]
pub struct Library {
    name: String,
    path: PathBuf,
}

pub const INJECT_LIBRARIES_RULE_NAME: &str = "inject_libraries";

/// A rule that removes trailing `nil` in local assignments.
#[derive(Debug, PartialEq, Eq)]
pub struct InjectLibraries {
    require_mode: RequireMode,
    libraries: Vec<Library>,
    path: PathBuf,
}

const DEFAULT_LIBRARIES_PATH: &str = "_DARKLUA_libs";

impl Default for InjectLibraries {
    fn default() -> Self {
        Self {
            require_mode: RequireMode::Path(Default::default()),
            libraries: Vec::new(),
            path: PathBuf::from_str(DEFAULT_LIBRARIES_PATH).unwrap(),
        }
    }
}

fn get_require_path(libs_path: &Path, path: &Path, context: &Context) -> PathBuf {
    let lib_file_stem = path
        .file_stem()
        .map(|x| OsString::from(x).into_string().unwrap())
        .unwrap();
    let lib_file_ext = path
        .extension()
        .map(|x| OsString::from(x).into_string().unwrap())
        .unwrap();

    let hash = blake3::hash(path.to_string_lossy().as_bytes());
    let hash_hex = hex::encode(&hash.as_bytes()[..8]);

    let lib_path = libs_path.join(format!("{}{}.{}", lib_file_stem, hash_hex, lib_file_ext));
    fs::copy(path, lib_path.as_path()).unwrap();
    let base_path = context
        .path
        .as_path()
        .parent()
        .expect("Could not find parent path of the source");
    let mut relative_path =
        diff_paths(lib_path.as_path(), base_path).expect("Could not resolve a path");
    relative_path.set_extension("");
    relative_path
}

impl Rule for InjectLibraries {
    fn process(&self, block: &mut Block, context: &Context) -> RuleProcessResult {
        let project_path = context
            .project_location
            .as_ref()
            .expect("Project path is required");
        let libs_path = project_path.join(self.path.as_path());
        fs::create_dir_all(&libs_path).unwrap();
        match self.require_mode.to_owned() {
            RequireMode::Path(_) => {
                for lib in &self.libraries {
                    let string_exp = StringExpression::from_value(
                        get_require_path(&libs_path, &lib.path, context).to_slash_lossy(),
                    );
                    let require_arg = Arguments::String(string_exp);

                    let require_call =
                        FunctionCall::new(Prefix::from_name("require"), require_arg, None);
                    let local_assignment = LocalAssignStatement::new(
                        vec![TypedIdentifier::new(lib.name.as_str())],
                        vec![require_call.into()],
                    );
                    block.insert_statement(0, local_assignment);
                }
            }
            RequireMode::Roblox(mut require_mode) => {
                require_mode
                    .initialize(context)
                    .map_err(|err| err.to_string())?;
                for lib in &self.libraries {
                    let require_path = get_require_path(&libs_path, &lib.path, context);
                    if let Some(require_arg) = require_mode
                        .generate_require(
                            require_path.as_path(),
                            &RequireMode::Path(PathRequireMode::new(require_path.to_slash_lossy())),
                            context,
                        )
                        .unwrap()
                    {
                        let require_call =
                            FunctionCall::new(Prefix::from_name("require"), require_arg, None);
                        let local_assignment = LocalAssignStatement::new(
                            vec![TypedIdentifier::new(lib.name.as_str())],
                            vec![require_call.into()],
                        );
                        block.insert_statement(0, local_assignment);
                    }
                }
            }
        }
        Ok(())
    }
}

impl RuleConfiguration for InjectLibraries {
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        verify_required_properties(&properties, &["require_mode", "libraries"])?;

        for (key, value) in properties {
            match key.as_str() {
                "require_mode" => {
                    self.require_mode = value.expect_require_mode(&key)?;
                }
                "libraries" => {
                    self.libraries = value.expect_libraries(&key)?;
                }
                _ => return Err(RuleConfigurationError::UnexpectedProperty(key)),
            }
        }

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        INJECT_LIBRARIES_RULE_NAME
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        RuleProperties::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::rules::Rule;

    use insta::assert_json_snapshot;

    fn new_rule() -> InjectLibraries {
        InjectLibraries::default()
    }

    #[test]
    fn serialize_default_rule() {
        let rule: Box<dyn Rule> = Box::new(new_rule());

        assert_json_snapshot!("default_inject_libraries", rule);
    }

    #[test]
    fn configure_with_extra_field_error() {
        let result = json5::from_str::<Box<dyn Rule>>(
            r#"{
            rule: "inject_libraries",
            require_mode: {
                name: "roblox"
            },
            libraries: [
                {
                    name: "task",
                    path: "task.luau"
                }
            ],
            prop: "something",
        }"#,
        );
        pretty_assertions::assert_eq!(result.unwrap_err().to_string(), "unexpected field 'prop'");
    }
}
