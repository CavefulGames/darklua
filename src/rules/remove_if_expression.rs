use crate::nodes::{
    self, BinaryExpression, Block, ElseIfExpressionBranch, Expression, IndexExpression,
    ParentheseExpression, TableEntry, TableExpression,
};
use crate::process::{DefaultVisitor, NodeProcessor, NodeVisitor};
use crate::rules::{
    Context, FlawlessRule, RuleConfiguration, RuleConfigurationError, RuleProperties,
};

use super::verify_no_rule_properties;

#[derive(Default)]
struct Processor {}

fn process(
    condition: &Expression,
    result: &Expression,
    else_result: &Expression,
    branches: &Vec<&ElseIfExpressionBranch>,
) -> Expression {
    let result_parenthese = ParentheseExpression::new(result.clone());
    let result_table = TableExpression::new(vec![TableEntry::Value(result_parenthese.into())]);
    let if_bin = BinaryExpression::new(nodes::BinaryOperator::And, condition.clone(), result_table);
    let bin_right = if branches.len() > 0 {
        let b = branches[0];
        process(
            b.get_condition(),
            b.get_result(),
            else_result,
            &branches[1..].to_vec(),
        )
    } else {
        let else_result_parenthese = ParentheseExpression::new(else_result.clone());
        let else_result_table =
            TableExpression::new(vec![TableEntry::Value(else_result_parenthese.into())]);
        else_result_table.into()
    };
    let bin = BinaryExpression::new(nodes::BinaryOperator::Or, if_bin, bin_right);
    let bin_parenthese = ParentheseExpression::new(Expression::Binary(Box::new(bin)));
    IndexExpression::new(bin_parenthese, 1).into()
}

impl NodeProcessor for Processor {
    fn process_expression(&mut self, expression: &mut Expression) {
        if let Expression::If(if_exp) = expression {
            let translated_exp = process(
                if_exp.get_condition(),
                if_exp.get_result(),
                if_exp.get_else_result(),
                &if_exp.iter_branches().collect(),
            );
            *expression = translated_exp;
        }

        // let call_exp: Option<Expression> = if let Expression::If(if_exp) = expression {
        //     let result_return = ReturnStatement::one(if_exp.get_result().clone());
        //     let else_result_return = ReturnStatement::one(if_exp.get_else_result().clone());
        //     let front_branch = IfBranch::new(if_exp.get_condition().clone(), result_return);
        //     let else_block = Block::new(vec![], Some(else_result_return.into()));

        //     let mut branches: Vec<IfBranch> = vec![front_branch];
        //     for elseif_exp in if_exp.iter_branches() {
        //         let elseif_result_return = ReturnStatement::one(elseif_exp.get_result().clone());
        //         let elseif_block = Block::new(vec![], Some(elseif_result_return.into()));
        //         branches.push(IfBranch::new(elseif_exp.get_condition().clone(), elseif_block));
        //     }

        //     let r#if = IfStatement::new(branches, Some(else_block));

        //     let func_block = Block::new(vec![r#if.into()], None);
        //     let func = Expression::Function(FunctionExpression::from_block(func_block));
        //     let parenthese_func = Prefix::Parenthese(ParentheseExpression::new(func));
        //     let func_call = FunctionCall::from_prefix(parenthese_func);
        //     Some(Expression::Call(Box::new(func_call)))
        // } else {
        //     None
        // };
        // if let Some(exp) = call_exp {
        //     *expression = exp;
        // }
    }
}

pub const REMOVE_IF_EXPRESSION_RULE_NAME: &str = "remove_if_expression";

/// A rule that removes trailing `nil` in local assignments.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct RemoveIfExpression {}

impl FlawlessRule for RemoveIfExpression {
    fn flawless_process(&self, block: &mut Block, _: &Context) {
        let mut processor = Processor::default();
        DefaultVisitor::visit_block(block, &mut processor);
    }
}

impl RuleConfiguration for RemoveIfExpression {
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        verify_no_rule_properties(&properties)?;

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        REMOVE_IF_EXPRESSION_RULE_NAME
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        RuleProperties::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::rules::Rule;

    use insta::assert_json_snapshot;

    fn new_rule() -> RemoveIfExpression {
        RemoveIfExpression::default()
    }

    #[test]
    fn serialize_default_rule() {
        let rule: Box<dyn Rule> = Box::new(new_rule());

        assert_json_snapshot!("default_remove_if_expression", rule);
    }

    #[test]
    fn configure_with_extra_field_error() {
        let result = json5::from_str::<Box<dyn Rule>>(
            r#"{
            rule: 'remove_if_expression',
            prop: "something",
        }"#,
        );
        pretty_assertions::assert_eq!(result.unwrap_err().to_string(), "unexpected field 'prop'");
    }
}
