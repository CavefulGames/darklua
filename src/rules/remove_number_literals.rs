use crate::{
    nodes::{Block, DecimalNumber, Expression, NumberExpression},
    process::{DefaultVisitor, NodeProcessor, NodeVisitor},
    rules::{Context, FlawlessRule, RuleConfiguration, RuleConfigurationError, RuleProperties},
};

use super::verify_no_rule_properties;

#[derive(Default)]
struct Processor {}

impl NodeProcessor for Processor {
    fn process_expression(&mut self, exp: &mut Expression) {
        if let Expression::Number(num_exp) = exp {
            match num_exp {
                NumberExpression::Binary(binary) => {
                    let value = binary.compute_value();
                    *exp = DecimalNumber::new(value).into();
                }
                NumberExpression::Decimal(decimal) => {
                    let value = decimal.compute_value();
                    *exp = DecimalNumber::new(value).into();
                }
                NumberExpression::Hex(hex) => {
                    let value = hex.compute_value();
                    *exp = DecimalNumber::new(value).into();
                }
            }
        }
    }
}

pub const REMOVE_NUMBER_LITERALS_RULE_NAME: &str = "remove_number_literals";

/// A rule that removes number literals.
#[derive(Default, Debug, PartialEq, Eq)]
pub struct RemoveNumberLiterals {}

impl FlawlessRule for RemoveNumberLiterals {
    fn flawless_process(&self, block: &mut Block, _: &Context) {
        let mut processor = Processor {};
        DefaultVisitor::visit_block(block, &mut processor);
    }
}

impl RuleConfiguration for RemoveNumberLiterals {
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        verify_no_rule_properties(&properties)?;

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        REMOVE_NUMBER_LITERALS_RULE_NAME
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        RuleProperties::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::rules::Rule;

    use insta::assert_json_snapshot;

    fn new_rule() -> RemoveNumberLiterals {
        RemoveNumberLiterals::default()
    }

    #[test]
    fn serialize_default_rule() {
        let rule: Box<dyn Rule> = Box::new(new_rule());

        assert_json_snapshot!("default_remove_number_literals", rule);
    }

    #[test]
    fn configure_with_extra_field_error() {
        let result = json5::from_str::<Box<dyn Rule>>(
            r#"{
            rule: 'remove_number_literals',
            prop: "something",
        }"#,
        );
        pretty_assertions::assert_eq!(result.unwrap_err().to_string(), "unexpected field 'prop'");
    }
}
