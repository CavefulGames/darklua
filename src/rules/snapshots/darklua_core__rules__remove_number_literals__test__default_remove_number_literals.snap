---
source: src/rules/remove_number_literals.rs
expression: rule
---
"remove_number_literals"
